File: tree/Traversal/Preorder/recursive/python/PreorderTraversal.py
from collections.abc import Sequence


def PreorderTraversal(left_children: Sequence[int], right_children: Sequence[int], root: int) -> list[int]:
    order: list[int] = []

    def dfs(node: int) -> None:
        order.append(node)

        left_child = left_children[node]
        if left_child != -1:
            dfs(left_child)

        right_child = right_children[node]
        if right_child != -1:
            dfs(right_child)

    dfs(root)
    return order

File: tree/Traversal/Preorder/recursive/cpp/PreorderTraversal.cpp
#include "PreorderTraversal.hpp"

#include <functional>
#include <vector>

using namespace std;

vector<int> PreorderTraversal(const vector<int>& leftChildren, const vector<int>& rightChildren, int root) {
    vector<int> order;

    function<void(int)> dfs = [&](int node) {
        order.push_back(node);

        if (leftChildren[node] != -1) {
            dfs(leftChildren[node]);
        }

        if (rightChildren[node] != -1) {
            dfs(rightChildren[node]);
        }
    };

    dfs(root);
    return order;
}
